### PR TITLE
Fix NPE in FIDFeatureProvider

### PR DESCRIPTION
**Pull request by sschulz92**
plus additional format fixes

Change-Id: Ib4b6856bb868dcaa572620a02855865f5c91e609

Signed-off-by: Sebastian Schulz <bastie92_spam@gmx.de>

### CODE REVIEW

File: plugins/org.locationtech.udig.tool.edit/src/org/locationtech/udig/tools/edit/behaviour/accept/AcceptChangesBehaviour.java [hunk 1]
```diff
@@ -263,20 +265,23 @@ public UndoableMapCommand getCommand(EditToolHandler handler) {
                     // commands.add(newFeatureCommand);
                     // }
                 } else {
-                    // not creating it so don't need to set it.
-                    UndoableMapCommand setGeometryCommand = new SetGeometryCommand(editGeom
-                            .getFeatureIDRef().get(), new StaticBlockingProvider<ILayer>(layer),
-                            SetGeometryCommand.DEFAULT, geom);
-                    commands.add(setGeometryCommand);
+                    String idRef = editGeom.getFeatureIDRef().get();
+                    if (idRef != null) {
```
Comment (fgdrf) [inline plugins/org.locationtech.udig.tool.edit/src/org/locationtech/udig/tools/edit/behaviour/accept/AcceptChangesBehaviour.java:269]: @sschulz92 I'm wondering about why the check is not implemented everywhere `editGeom.getFeatureIDRef().get()` occurs within this class.

I would expect to check in the other else-block down below as well. Therfore I think its worth to assign it to a new variable once and use it in both blocks to verify of none null.

**Review by fgdrf — CHANGES_REQUESTED** [review]
Thanks for this pull request. I'm wondering if you have a stack-trace or can describe when the problem occurs.

**Comment by sschulz92** [comment]
The related stacktrace looks like:

```
java.lang.NullPointerException: Fid must not be null
	at org.locationtech.udig.project.command.provider.FIDFeatureProvider.<init>(FIDFeatureProvider.java:33)
	at org.locationtech.udig.project.internal.commands.edit.SetAttributeCommand.<init>(SetAttributeCommand.java:78)
	at org.locationtech.udig.project.internal.commands.edit.SetGeometryCommand.<init>(SetGeometryCommand.java:68)
	at org.locationtech.udig.tools.edit.behaviour.accept.AcceptChangesBehaviour.processIntoCommands(AcceptChangesBehaviour.java:263)
	at org.locationtech.udig.tools.edit.behaviour.accept.AcceptChangesBehaviour.getCommand(AcceptChangesBehaviour.java:164)
	at org.locationtech.udig.tools.edit.BehaviourCommand.execute(BehaviourCommand.java:53)
	at org.locationtech.udig.project.command.UndoableComposite.execute(UndoableComposite.java:79)
	at org.locationtech.udig.project.command.CommandManager$Executor.execute(CommandManager.java:395)
	at org.locationtech.udig.project.command.CommandManager$Executor.run(CommandManager.java:326)
	at org.locationtech.udig.project.command.CommandManager$Executor.run(CommandManager.java:312)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

According to your review request I will provide another commit soon :)

File: plugins/org.locationtech.udig.tool.edit/src/org/locationtech/udig/tools/edit/behaviour/accept/AcceptChangesBehaviour.java (hunk 1 above)
Comment (sschulz92) [reply plugins/org.locationtech.udig.tool.edit/src/org/locationtech/udig/tools/edit/behaviour/accept/AcceptChangesBehaviour.java]: Can you have a second look please? I just pushed the change in order to fulfill your review.

File: plugins/org.locationtech.udig.tool.edit/src/org/locationtech/udig/tools/edit/behaviour/accept/AcceptChangesBehaviour.java [hunk 2]
```diff
@@ -265,19 +267,20 @@ public UndoableMapCommand getCommand(EditToolHandler handler) {
...
                                 geom);
                         commands.add(setGeometryCommand);
                     }
                 }
             } else {
-                commands.add(new CreateNewOrSelectExitingFeatureCommand(
-                        editGeom.getFeatureIDRef().get(), layer, geom));
-
+                if (featureIdReference != null) {
+                    CreateNewOrSelectExitingFeatureCommand featureCommand = new CreateNewOrSelectExitingFeatureCommand(
```
Comment (sschulz92) [inline plugins/org.locationtech.udig.tool.edit/src/org/locationtech/udig/tools/edit/behaviour/accept/AcceptChangesBehaviour.java:280]: @fgdrf may I ask if I just found another typo here? As far as I can see it should be "CreateNewOrSelectExistingFeatureCommand" right?

If yes, I will rename the class after this PR is merged.

Comment (fgdrf) [reply plugins/org.locationtech.udig.tool.edit/src/org/locationtech/udig/tools/edit/behaviour/accept/AcceptChangesBehaviour.java:280]: Good catch, we can adress it in another changest ;)

**Review by fgdrf — COMMENTED** [review]
> ```
> java.lang.NullPointerException: Fid must not be null
> 	at org.locationtech.udig.project.command.provider.FIDFeatureProvider.<init>(FIDFeatureProvider.java:33)
> 	at org.locationtech.udig.project.internal.commands.edit.SetAttributeCommand.<init>(SetAttributeCommand.java:78)
> 	at org.locationtech.udig.project.internal.commands.edit.SetGeometryCommand.<init>(SetGeometryCommand.java:68)
> 	at org.locationtech.udig.tools.edit.behaviour.accept.AcceptChangesBehaviour.processIntoCommands(AcceptChangesBehaviour.java:263)
> 	at org.locationtech.udig.tools.edit.behaviour.accept.AcceptChangesBehaviour.getCommand(AcceptChangesBehaviour.java:164)
> 	at org.locationtech.udig.tools.edit.BehaviourCommand.execute(BehaviourCommand.java:53)
> 	at org.locationtech.udig.project.command.UndoableComposite.execute(UndoableComposite.java:79)
> 	at org.locationtech.udig.project.command.CommandManager$Executor.execute(CommandManager.java:395)
> 	at org.locationtech.udig.project.command.CommandManager$Executor.run(CommandManager.java:326)
> 	at org.locationtech.udig.project.command.CommandManager$Executor.run(CommandManager.java:312)
> 	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
> ```

Great, it helps to understand. Do you remember the workflow how to reproduce the error?

**Comment by sschulz92** [comment]
While I investigated a little bit I found the following information:

1. Open view 'Map' and select the Create Polygon Geometries tool to draw on the map.
2. Draw an invalid polygon geometry.
3. Close the Invalid Geometry pop-up.
4. Right click (context menu) on the polygon geometry and select the delete option.
5. Check the Error Log view of the ICAS-client for the NullPointerException.

Note: internally I am using a ViewPart instead of an EditorPart, therefore this NPE might not be reproduceable using plain uDig.

**Comment by fgdrf** [comment]
> Note: internally I am using a ViewPart instead of an EditorPart, therefore this NPE might not be reproduceable using plain uDig.

Thanks for this description, I can check it with the current state of uDig and if the patchset helps here :)

**Comment by fgdrf** [comment]
I could reproduce the Problem with the current Application build from master-branch (https://github.com/locationtech/udig-platform/runs/3633771229)

Following the workflow description I still get an exception thrown with artefact from build for this pull-request:

```
java.lang.NullPointerException: Fid must not be null
	at org.locationtech.udig.project.command.provider.FIDFeatureProvider.<init>(FIDFeatureProvider.java:41)
	at org.locationtech.udig.tools.edit.behaviour.accept.AcceptChangesBehaviour.processIntoCommands(AcceptChangesBehaviour.java:216)
	at org.locationtech.udig.tools.edit.behaviour.accept.AcceptChangesBehaviour.getCommand(AcceptChangesBehaviour.java:169)
	at org.locationtech.udig.tools.edit.BehaviourCommand.execute(BehaviourCommand.java:53)
	at org.locationtech.udig.project.command.UndoableComposite.execute(UndoableComposite.java:79)
	at org.locationtech.udig.project.command.CommandManager$Executor.execute(CommandManager.java:397)
	at org.locationtech.udig.project.command.CommandManager$Executor.run(CommandManager.java:328)
	at org.locationtech.udig.project.command.CommandManager$Executor.run(CommandManager.java:314)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
``` 

I assume the `entry.getKey()` is null ..
```
 FIDFeatureProvider featureProvider = new FIDFeatureProvider(entry.getKey(),
                    layerProvider);
```

**Comment by sschulz92** [comment]
@fgdrf the latest commit fixed the NullPointerException finally :) Can you have a last look onto the PR please?

**Comment by fgdrf** [comment]
:heavy_check_mark:  tested again and don't see exception anymore.

**Pull request merged**